source/vecs/storage.d

```diff
@@ -461,34 +470,30 @@ package:
 
 
 private:
-	/**
-	 * Associates a component to an entity. Emits onSet after associating the
-	 *     component to the entity.
-	 *
-	 * Safety: The **internal code** is @safe, however, beacause of **Signal**
-	 *     dependency, the method must be @system.
-	 *
-	 * Signal: Emits onSet **after** the component is set.
-	 *
-	 * Returns: `Component*` pointing to the component set.
-	 */
-	@system
-	Component* _set(in Entity entity, Component component)
+	// FIXME: documentation
+	Component* _add(in Entity e)
+		in (!(e.id < _sparsedEntities.length
+			&& _sparsedEntities[e.id] < _packedEntities.length
+			&& _packedEntities[_sparsedEntities[e.id]].id == e.id
+			&& _packedEntities[_sparsedEntities[e.id]].batch != e.batch
+		))
 	{
-		_packedEntities ~= entity; // set entity
-		_components ~= component; // set component
+		if (!contains(e))
+		{
+			_packedEntities ~= e; // set entity
+			_components.length++;
 
-		// map to the correct entity from the packedEntities from sparsedEntities
-		if (entity.id >= _sparsedEntities.length) _sparsedEntities.length = entity.id + 1;
-			_sparsedEntities[entity.id] = _packedEntities.length - 1; // safe cast
+			// map to the correct entity from the packedEntities from sparsedEntities
+			if (e.id >= _sparsedEntities.length)
+				_sparsedEntities.length = e.id + 1;
```

Comment:
Are you aware of GC allocations here? Doesn't this trigger a huge number of allocations when constantly adding entities? If yes, I would suggest taking the approach of std.array.Appender to reduce that.

Reply:
This is done once only. I don't have functions that can add multiple entities to a component at once.

Reply:
> This is done once only. I don't have functions that can add multiple entities to a component at once.

Oh ok :+1:

Reply:
I can add in the future a function to `reserve` memory. This is what `Appender` uses internally. The array has a built-in `reserve` function to reserve memory without initializing it, and this is what increases the performance versus the standard appending way.